poll/obs/get.js

```diff
@@ -0,0 +1,189 @@
+//
+// Things that are not observable and can all live in one Value which will get set just once.:
+//  - title, author, body, channel, mentions, recps, the original poll message.
+//  {
+//    sync: boolean, (initially false until the value gets set)
+//    poll: MutantValue({
+//      title: '',
+//      author: '',
+//      body: '',
+//      channel: '',
+//      mentions: '',
+//      recps: '',
+//      value: {...poll msg}
+//    }),
+//    closesAt: MutantValue('date string'),
+//    positions: MutantArray(), (sorted in causal order)
+//    results: MutantStruct(),A computed obs of positions
+//    errors
+//  }
+//  The shape of this object is different to the one returned by async.get. Ugh.
+
+module.exports = function (server) {
+  return function get (key) {
+    const myKey = server.id
+
+    const positions = MutantArray([])
+    const myPositions = MutantArray([])
+    const closingTimes = MutantArray([])
+    const sortedClosingTimes = computed(closingTimes, sort)
+    const sortedPositions = computed(positions, sort)
+    const closesAt = computed(sortedClosingTimes, (times) => {
+      const time = times.pop()
+      return time ? time.value.content.closesAt : ''
+    })
+    const myPosition = computed(myPositions, (positions) => {
+      const sortedPositions = sort(positions)
+      return sortedPositions.pop()
+    })
+
+    const poll = Value({})
+    const results = computed(sortedPositions, (positions) => {
+      const resultsErrors = buildResults({ poll: resolve(poll), positions })
+      return resultsErrors ? resultsErrors.results : {}
+    })
+
+    const errors = computed(sortedPositions, (positions) => {
+      const resultsErrors = buildResults({ poll: resolve(poll), positions })
+      return resultsErrors ? resultsErrors.errors : {}
+    })
+
+    const pollDoc = Struct({
+      sync: false,
+      positions: sortedPositions,
+      closesAt,
+      results,
+      errors,
+      poll,
+      myPosition
+    })
+
+    server.get(key, (err, value) => {
+      if (err) return err
+
+      var poll = { key, value }
+      if (!isPoll(poll)) return new Error('scuttle-poll could not get poll, key provided was not a valid poll key')
+
+      // give subscribers a chance to start listening so they don't miss updates.
+      setImmediate(function () {
+        pollDoc.poll.set(decoratePoll(poll))
+
+        const refs = PullNotify()
+
+        pull(
+          createBacklinkStream(key),
+          pull.drain(refs)
+        )
+
+        // don't sync obs until we got sync from the stream to save some renders.
+        pull(
+          refs.listen(),
+          pull.filter(ref => ref.sync),
+          pull.drain(() => {
+            // allow the other streams to update their observables before sync goes true
+            setImmediate(() => pollDoc.sync.set(true))
+          })
+        )
+
+        pull(
+          refs.listen(),
+          pull.filter(isPosition[CHOOSE_ONE]), // TODO: this shouldn't be hard coded
```

Comment:
get rid of CHOOSE_ONE by testing the poll we get is a valid poll, then testing what sort of poll it is.
Might be nice to have a method (in ssb-poll-schema?)

```js

function getPollType (msg) {
  if (!isPoll(msg)) return

  if (isChooseOnePoll(msg)) return CHOOSE_ONE
  if (isChooseDotPoll(msg)) return DOT
  //...

}

Reply:
Yes, great, was just thinking about adding this to poll-schema before.